velox/functions/lib/tests/TDigestTest.cpp

```diff
@@ -0,0 +1,384 @@
+  std::uniform_real_distribution<> dist;
+  for (int i = 0; i < N; ++i) {
+    auto v = dist(gen);
+    digest.add(positions, v);
+    values[i] = v;
+  }
+  digest.compress(positions);
+  std::sort(std::begin(values), std::end(values));
+  CHECK_QUANTILES(folly::Range(values, N), digest);
+}
+
+TEST(TDigestTest, fewElements) {
+  TDigest digest;
+  std::vector<int16_t> positions;
+  digest.compress(positions);
+  ASSERT_EQ(digest.sum(), 0);
+  for (auto q : kQuantiles) {
+    ASSERT_TRUE(std::isnan(digest.estimateQuantile(q)));
+  }
+  digest.add(positions, 1.0);
+  digest.compress(positions);
+  ASSERT_EQ(digest.sum(), 1);
+  for (auto q : kQuantiles) {
+    ASSERT_EQ(digest.estimateQuantile(q), 1.0);
+  }
+}
+
+// IMPORTANT: All these errors cannot be caught by TRY in Presto, so we should
+// not make them user errors.  If in another engine these are catchable errors,
+// throw user errors in the corresponding UDFs before they reach the TDigest
+// implementation.
+TEST(TDigestTest, invalid) {
+  TDigest digest;
+  ASSERT_THROW(digest.setCompression(NAN), VeloxRuntimeError);
+  ASSERT_THROW(digest.setCompression(0), VeloxRuntimeError);
+  ASSERT_THROW(digest.setCompression(1000.1), VeloxRuntimeError);
+  std::vector<int16_t> positions;
+  ASSERT_THROW(digest.add(positions, NAN), VeloxRuntimeError);
+  ASSERT_THROW(digest.add(positions, 1, 0), VeloxRuntimeError);
+  ASSERT_THROW(digest.estimateQuantile(1.1), VeloxRuntimeError);
+}
+
+TEST(TDigestTest, unalignedSerialization) {
+  constexpr int N = 1e4;
+  TDigest digest;
+  std::vector<int16_t> positions;
+  for (int i = 0; i < N; ++i) {
+    digest.add(positions, i);
+  }
+  digest.compress(positions);
+  ASSERT_NEAR(digest.sum(), 1.0 * N * (N - 1) / 2, kSumError);
+  std::string buf(1 + digest.serializedByteSize(), '\0');
+  for (int offset = 0; offset < 2; ++offset) {
+    SCOPED_TRACE(fmt::format("offset={}", offset));
+    digest.serialize(buf.data() + offset);
+    TDigest digest2;
+    digest2.mergeDeserialized(positions, buf.data() + offset);
+    digest2.compress(positions);
+    for (auto q : kQuantiles) {
+      auto v = digest2.estimateQuantile(q);
+      ASSERT_NEAR(v / (N - 1), q, kRankError);
+    }
+  }
+}
+
+TEST(TDigestTest, mergeEmpty) {
+  std::vector<int16_t> positions;
+  TDigest<> digests[2];
+  std::string buf(digests[1].serializedByteSize(), '\0');
+  digests[1].serialize(buf.data());
+  digests[0].mergeDeserialized(positions, buf.data());
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 0);
+  ASSERT_TRUE(std::isnan(digests[0].estimateQuantile(0.5)));
+  digests[0].add(positions, 1.0);
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 1);
+  ASSERT_EQ(digests[0].estimateQuantile(0.5), 1);
+  digests[0].mergeDeserialized(positions, buf.data());
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 1);
+  ASSERT_EQ(digests[0].estimateQuantile(0.5), 1);
+}
+
+TEST(TDigestTest, deserializeJava) {
```

Comment:
Could we also add a case where the C++ T-Digest already contains some values  and then merged with a Java one (both Java merged into C++ and C++ merged into Java)?

Also, what happens if a C++ T-Digest hasn't been compressed when we merge a Java one into it?

Reply:
Another question, should we also test the SerDe of T-Digest of the old version? And merging old version with new version of T-Digests?

Reply:
We can add a `mergeWithJava` for sanity check, the code path is exactly the same as merging C++ one so we don't need very sophisticated test on that.

The only difference between version 0 and version 1 is the existence of sum so I don't think we need separate test for it.  We don't know how to get version 0 data either.

Reply:
@Yuhta I see. Since kSerializationVersion is hard-coded as 1, the C++ T-Digest will never be serialized with version 0. It would be ideal if we can get Java-serialized version 0, but since we don't know how to get that, it's OK to leave it out. (Could you double check with Presto folks whether they have an easy way to produce version-0 T-Digest, such as through a query config?)

Reply:
I don't think there is a way to write version 0 in current Java version, it's hard-coded to 1: https://github.com/prestodb/presto/blob/3b1aff96e3d6bcc9483a4aabe1aa3fdaf380b309/presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java#L718